core/includes/form.inc

```diff
@@ -2975,14 +2977,29 @@ function theme_fieldset($variables) {
     $output .= '<legend><span class="fieldset-legend">' . $element['#title'] . '</span></legend>';
   }
   $output .= '<div class="fieldset-wrapper">';
+
+  $description_before = '';
+  $description_after = '';
   if (!empty($element['#description'])) {
-    $output .= '<div class="fieldset-description">' . $element['#description'] . '</div>';
+    $description = theme('form_element_description', $variables);
+
+    switch ($element['#description_display']) {
+      case 'before':
+        $description_before = $description;
+        break;
+
+      case 'after':
+      case 'invisible':
+        $description_after = $description;
+        break;
+    }
   }
-  $output .= $element['#children'];
+
+  $output .= $description_before . $element['#children'];
```

Comment:
This is totally minor but I usually like to separate concatenations if they're discrete HTML elements:

```suggestion
  $output .= $description_before;
  $output .= $element['#children'];
```